src/piton/jax.py

```diff
@@ -0,0 +1,1095 @@
+    b is a 2-d tensor of shape (B, K)
+
+    indices is a 2-d tensor of shape (N, 2), such that indices[:, 0] are indices into A and indices[:, 1] are indices into B.
+
+    This returns a (N,) shaped tensor that contains the dot products for each row in indices.
+    """
+    return embedding_dot_fwd(a, b, indices)[0]
+
+
+def embedding_dot_fallback(a: Array, b: Array, indices: Array):
+    """The fallback implementation for embedding_dot."""
+    a_indices = indices[:, 0]
+    b_indices = indices[:, 1]
+
+    a_vals = jnp.take(
+        a,
+        a_indices,
+        axis=0,
+        mode="fill",
+        fill_value=0,
+    )
+
+    b_vals = jnp.take(
+        b,
+        b_indices,
+        axis=0,
+        mode="fill",
+        fill_value=0,
+    )
+
+    return jnp.multiply(a_vals, b_vals).sum(axis=1)
+
+
+def embedding_dot_fwd(
+    a: Array, b: Array, indices: Array
+) -> Tuple[Array, Tuple[Array, Array, Array]]:
+    """The forward pass for embedding dot."""
+    r = embedding_dot_forward_p.bind(a, b, indices)
+    return r, (a, b, indices)
+
+
+def embedding_dot_bwd(
+    res: Tuple[Array, Array, Array], g: Array
+) -> Tuple[Array, Array, None]:
+    """The backward pass for embedding dot."""
+    a, b, indices = res
+    da, db = embedding_dot_backward_p.bind(a, b, indices, g)
+    return (da, db, None)
+
+
+embedding_dot.defvjp(embedding_dot_fwd, embedding_dot_bwd)
+
+embedding_dot_forward_p = jax.core.Primitive("embedding_dot_forward")
+
+
+def embedding_dot_forward_p_abstract_eval(
+    a: jax.core.ShapedArray,
+    b: jax.core.ShapedArray,
+    indices: jax.core.ShapedArray,
+):
+    """The abstract shape for the forward primitive."""
+    assert a.dtype == b.dtype
+    assert indices.dtype == jnp.uint32
+    return jax.core.ShapedArray((indices.shape[0],), a.dtype)
+
+
+def embedding_dot_forward_p_xla_translation(
+    ctx: jax.interpreters.xla.TranslationContext,
+    avals_in: Sequence[jax.core.AbstractValue],
+    avals_out: Sequence[jax.core.AbstractValue],
+    a: jax.interpreters.xla.XlaOp,
+    b: jax.interpreters.xla.XlaOp,
+    indices: jax.interpreters.xla.XlaOp,
+) -> Sequence[jax.interpreters.xla.XlaOp]:
+    """The actual hardware implementation of embedding_dot_forward_p."""
+    avals_in = _assert_and_cast_to_shaped(avals_in)
+    avals_out = _assert_and_cast_to_shaped(avals_out)
+
+    a_shape, b_shape, indices_shape = avals_in
+    (result_shape,) = avals_out
+
+    if ctx.platform == "cpu":
+        computation = jax.xla_computation(embedding_dot_fallback)(*avals_in)
+        res = xla_client.ops.Call(ctx.builder, computation, [a, b, indices])
+        return [
+            xla_client.ops.GetTupleElement(res, i)
+            for i, _ in enumerate(avals_out)
+        ]
+    elif ctx.platform == "cuda":
+        assert a_shape.dtype == b_shape.dtype
+
+        if a_shape.dtype == np.float32:
+            name = "float_embedding_dot_forward"
+        elif a_shape.dtype == np.float16:
```

Comment:
Because I only added dtype support in a lazy fashion. Adding only when I needed to. 

These functions should in principle support every dtype, it's just a question of the time to add that support.